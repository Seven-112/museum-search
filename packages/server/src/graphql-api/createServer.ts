import { ApolloServer } from "apollo-server";
import { Client } from "elasticsearch";
import { resolvers } from "./resolvers";
import { typeDefs } from "./typeDefs";
import { IResolverContext } from "./types";

interface ICreateServerParams {
  esClient: Client;
}

export function createServer({ esClient }: ICreateServerParams) {
  return new ApolloServer({
    context: { esClient } as IResolverContext,
    resolvers,
    typeDefs
  });
}
