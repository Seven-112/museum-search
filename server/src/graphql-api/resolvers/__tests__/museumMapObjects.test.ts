import { gql } from "apollo-server";
import { createTestClient } from "apollo-server-testing";
import { Client, SearchParams } from "elasticsearch";
import { createServer } from "../../createServer";

const MUSEUM_MAP_OBJECT_QUERY = gql`
  query museumMapObjects($query: String, $boundingBox: GeoBoundingBoxInput) {
    museumMapObjects(query: $query, boundingBox: $boundingBox) {
      edges {
        ... on MuseumSearchEdge {
          node {
            __typename
            id
            name
            latitude
            longitude
          }
        }
        ... on GeoPointBucketEdge {
          node {
            __typename
            geoHashKey
            latitude
            longitude
            count
          }
        }
      }
    }
  }
`;

const MUSEUM_BUCKETS_MOCK_RESPONSE = {
  aggregations: {
    museumsGrid: {
      buckets: [
        {
          key: "dr4",
          doc_count: 931,
          avgLongitude: {
            value: -75.22248325696181
          },
          avgLatitude: {
            value: 40.11064952682348
          }
        },
        {
          key: "dr7",
          doc_count: 788,
          avgLongitude: {
            value: -73.81607840872053
          },
          avgLatitude: {
            value: 41.196299010727
          }
        },
        {
          key: "dr5",
          doc_count: 2,
          avgLongitude: {
            value: -74.01199078959803
          },
          avgLatitude: {
            value: 40.610638550789126
          }
        }
      ]
    }
  }
};

const MUSEUMS_MOCK_RESPONSE = {
  hits: {
    hits: [
      {
        _index: "museums",
        _type: "museum",
        _id: "5104",
        _score: 0,
        _source: {
          id: 5104,
          name: "OPERATION SAIL",
          legalName: "OPERATION SAIL INC",
          alternateName: null,
          museumType: "HISTORIC PRESERVATION",
          institutionName: null,
          streetAddress: "55 WATER ST",
          city: "NEW YORK",
          state: "NY",
          zipCode: "6902",
          phoneNumber: "2124222162",
          latitude: 40.70317,
          longitude: -74.00916,
          createdAt: "2018-11-25T22:48:02.000Z",
          updatedAt: "2018-11-25T22:48:02.000Z",
          location: {
            lat: 40.70317,
            lon: -74.00916
          }
        }
      },
      {
        _index: "museums",
        _type: "museum",
        _id: "18747",
        _score: 0,
        _source: {
          id: 18747,
          name: "ABRAHAM STAATS HOUSE",
          legalName: "FRIENDS OF ABRAHAM STAATS HOUSE INC",
          alternateName: null,
          museumType: "HISTORY MUSEUM",
          institutionName: null,
          streetAddress: null,
          city: null,
          state: null,
          zipCode: null,
          phoneNumber: "7324695836",
          latitude: 40.5535,
          longitude: -74.52932,
          createdAt: "2018-11-25T22:48:02.000Z",
          updatedAt: "2018-11-25T22:48:02.000Z",
          location: {
            lat: 40.5535,
            lon: -74.52932
          }
        }
      }
    ]
  }
};

jest.mock("elasticsearch", () => ({
  Client: class {
    async search(params: SearchParams) {
      if (params.body.aggregations) {
        return MUSEUM_BUCKETS_MOCK_RESPONSE;
      } else {
        return MUSEUMS_MOCK_RESPONSE;
      }
    }
  }
}));

const { query } = createTestClient(createServer({ esClient: new Client({}) }));

describe("museumMapObjects resolver", () => {
  it("returns a MuseumMapObjectsConnection when no query string or bounding box is provided", async () => {
    // Spy on the Elasticsearch client's "search" method.
    const search = jest.spyOn(
      require("elasticsearch").Client.prototype,
      "search"
    );

    // Do the query.
    const response = await query({ query: MUSEUM_MAP_OBJECT_QUERY });

    // Check that the client search method was called twice with the correct args.
    expect(search).toBeCalledTimes(2);
    const [bucketsCall, museumsCall] = search.mock.calls;
    expect(bucketsCall).toMatchSnapshot("Buckets search with no arguments.");
    expect(museumsCall).toMatchSnapshot("Museums search with no arguments.");

    // Check that the GraphQL response is correct.
    expect(response).toMatchSnapshot("museumMapObjects with no args response.");
  });
});
