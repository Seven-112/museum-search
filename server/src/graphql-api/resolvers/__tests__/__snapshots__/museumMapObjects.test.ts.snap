// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection when a query string and bounding box are provided: Buckets search with query string and bounding box args. 1`] = `
Array [
  Object {
    "body": Object {
      "aggregations": Object {
        "museumsGrid": Object {
          "aggregations": Object {
            "avgLatitude": Object {
              "avg": Object {
                "field": "latitude",
              },
            },
            "avgLongitude": Object {
              "avg": Object {
                "field": "longitude",
              },
            },
          },
          "geohash_grid": Object {
            "field": "location",
            "precision": 3,
          },
        },
      },
      "query": Object {
        "bool": Object {
          "filter": Object {
            "geo_bounding_box": Object {
              "location": Object {
                "bottom": -0.7031073524364783,
                "left": -150.20489340321566,
                "right": -45.61504965321564,
                "top": 65.14611484756375,
              },
            },
          },
          "must": Object {
            "multi_match": Object {
              "query": "museum",
            },
          },
        },
      },
    },
    "index": "museums",
    "size": 0,
  },
]
`;

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection when a query string and bounding box are provided: Museums search with query string and bounding box args. 1`] = `
Array [
  Object {
    "body": Object {
      "query": Object {
        "bool": Object {
          "minimum_should_match": 1,
          "must": Object {
            "multi_match": Object {
              "query": "museum",
            },
          },
          "should": Array [
            Object {
              "bool": Object {
                "filter": Object {
                  "geo_bounding_box": Object {
                    "location": Object {
                      "bottom": 39.375,
                      "left": -74.53125,
                      "right": -73.125,
                      "top": 40.78125,
                    },
                  },
                },
              },
            },
          ],
        },
      },
    },
    "index": "museums",
    "size": 5000,
  },
]
`;

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection when a query string and bounding box are provided: museumMapObjects with query string and bounding box args response. 1`] = `
Object {
  "data": Object {
    "museumMapObjects": Object {
      "edges": Array [
        Object {
          "node": Object {
            "__typename": "GeoPointBucket",
            "count": 931,
            "geoHashKey": "dr4",
            "latitude": 40.11064952682348,
            "longitude": -75.22248325696181,
          },
        },
        Object {
          "node": Object {
            "__typename": "GeoPointBucket",
            "count": 788,
            "geoHashKey": "dr7",
            "latitude": 41.196299010727,
            "longitude": -73.81607840872053,
          },
        },
        Object {
          "node": Object {
            "__typename": "Museum",
            "id": 5104,
            "latitude": 40.70317,
            "longitude": -74.00916,
            "name": "OPERATION SAIL",
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection when no query string or bounding box is provided: Buckets search with no arguments. 1`] = `
Array [
  Object {
    "body": Object {
      "aggregations": Object {
        "museumsGrid": Object {
          "aggregations": Object {
            "avgLatitude": Object {
              "avg": Object {
                "field": "latitude",
              },
            },
            "avgLongitude": Object {
              "avg": Object {
                "field": "longitude",
              },
            },
          },
          "geohash_grid": Object {
            "field": "location",
            "precision": 2,
          },
        },
      },
      "query": Object {
        "bool": Object {
          "filter": undefined,
          "must": undefined,
        },
      },
    },
    "index": "museums",
    "size": 0,
  },
]
`;

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection when no query string or bounding box is provided: Museums search with no arguments. 1`] = `
Array [
  Object {
    "body": Object {
      "query": Object {
        "bool": Object {
          "minimum_should_match": 1,
          "must": undefined,
          "should": Array [
            Object {
              "bool": Object {
                "filter": Object {
                  "geo_bounding_box": Object {
                    "location": Object {
                      "bottom": 39.375,
                      "left": -74.53125,
                      "right": -73.125,
                      "top": 40.78125,
                    },
                  },
                },
              },
            },
          ],
        },
      },
    },
    "index": "museums",
    "size": 5000,
  },
]
`;

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection when no query string or bounding box is provided: museumMapObjects with no args response. 1`] = `
Object {
  "data": Object {
    "museumMapObjects": Object {
      "edges": Array [
        Object {
          "node": Object {
            "__typename": "GeoPointBucket",
            "count": 931,
            "geoHashKey": "dr4",
            "latitude": 40.11064952682348,
            "longitude": -75.22248325696181,
          },
        },
        Object {
          "node": Object {
            "__typename": "GeoPointBucket",
            "count": 788,
            "geoHashKey": "dr7",
            "latitude": 41.196299010727,
            "longitude": -73.81607840872053,
          },
        },
        Object {
          "node": Object {
            "__typename": "Museum",
            "id": 5104,
            "latitude": 40.70317,
            "longitude": -74.00916,
            "name": "OPERATION SAIL",
          },
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection with no edges when an empty bounding box is provided: Buckets search with empty bounding box. 1`] = `
Array [
  Object {
    "body": Object {
      "aggregations": Object {
        "museumsGrid": Object {
          "aggregations": Object {
            "avgLatitude": Object {
              "avg": Object {
                "field": "latitude",
              },
            },
            "avgLongitude": Object {
              "avg": Object {
                "field": "longitude",
              },
            },
          },
          "geohash_grid": Object {
            "field": "location",
            "precision": 12,
          },
        },
      },
      "query": Object {
        "bool": Object {
          "filter": Object {
            "geo_bounding_box": Object {
              "location": Object {
                "bottom": 38.907974370742046,
                "left": -94.47119146585467,
                "right": -94.46789234876636,
                "top": 38.909752622714755,
              },
            },
          },
          "must": undefined,
        },
      },
    },
    "index": "museums",
    "size": 0,
  },
]
`;

exports[`museumMapObjects resolver returns a MuseumMapObjectsConnection with no edges when an empty bounding box is provided: museumMapObjects with query string and bounding box args response. 1`] = `
Object {
  "data": Object {
    "museumMapObjects": Object {
      "edges": Array [],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
